Source/WebCore/platform/graphics/mac/controls/ControlMac.mm

```diff
@@ -176,6 +177,43 @@
     [WebControlWindow setHasKeyAppearance:style.states.contains(ControlStyle::State::WindowActive)];
 }
 
+static void drawViewlessCell(const FloatRect& rect, float deviceScaleFactor, const ControlStyle& style, NSCell *cell, bool flipContext)
+{
+    CGContextRef cgContext = [[NSGraphicsContext currentContext] CGContext];
+    CGContextStateSaver stateSaver(cgContext);
+
+    [cell _setFallbackBackingScaleFactor:deviceScaleFactor];
+
+#if USE(NSVIEW_SEMANTICCONTEXT)
+    if (style.states.contains(ControlStyle::State::FormSemanticContext))
+        [cell _setFallbackSemanticContext:NSViewSemanticContextForm];
+#endif
+
+    auto adjustedRect = rect;
+
+    // FIXME: Remove flipping the coordinates if rdar://105194487 is fixed.
+    if (flipContext) {
+        // Move the coordinates origin to topleft of rect.
+        CGContextTranslateCTM(cgContext, adjustedRect.x(), adjustedRect.y());
+        adjustedRect.setLocation(FloatPoint::zero());
+
+        // Flip the coordinates.
+        CGContextTranslateCTM(cgContext, 0, adjustedRect.height());
+        CGContextScaleCTM(cgContext, 1, -1);
+    }
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wnonnull"
+    [cell drawWithFrame:adjustedRect inView:nil];
+#pragma clang diagnostic pop
+}
+
+static void drawCellInView(const FloatRect& rect, NSCell *cell, NSView *view)
+{
+    [cell drawWithFrame:rect inView:view];
+    [cell setControlView:nil];
+}
+
 static void drawFocusRing(NSRect cellFrame, NSCell *cell, NSView *view)
```

Comment:
Do we need a viewless approach here as well?

Reply:
Yes we do. But when passing a nil NSView to [cell drawFocusRingMaskWithFrame:...] nothing is displayed. I added a FIXME comment and I filed rdar://105249508.